coderd/notifications/notifier.go

```diff
@@ -0,0 +1,247 @@
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("acquire messages: %w", err)
+	}
+
+	return msgs, nil
+}
+
+// prepare has two roles:
+// 1. render the title & body templates
+// 2. build a dispatcher from the given message, payload, and these templates - to be used for delivering the notification
+func (n *notifier) prepare(ctx context.Context, msg database.AcquireNotificationMessagesRow) (dispatch.DeliveryFunc, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	// NOTE: when we change the format of the MessagePayload, we have to bump its version and handle unmarshalling
+	// differently here based on that version.
+	var payload types.MessagePayload
+	err := json.Unmarshal(msg.Payload, &payload)
+	if err != nil {
+		return nil, xerrors.Errorf("unmarshal payload: %w", err)
+	}
+
+	handler, ok := n.handlers[msg.Method]
+	if !ok {
+		return nil, xerrors.Errorf("failed to resolve handler %q", msg.Method)
+	}
+
+	var title, body string
+	if title, err = render.GoTemplate(msg.TitleTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render title: %w", err)
+	}
+	if body, err = render.GoTemplate(msg.BodyTemplate, payload, nil); err != nil {
+		return nil, xerrors.Errorf("render body: %w", err)
+	}
+
+	return handler.Dispatcher(payload, title, body)
+}
+
+// deliver sends a given notification message via its defined method.
+// This method *only* returns an error when a context error occurs; any other error is interpreted as a failure to
+// deliver the notification and as such the message will be marked as failed (to later be optionally retried).
+func (n *notifier) deliver(ctx context.Context, msg database.AcquireNotificationMessagesRow, deliver dispatch.DeliveryFunc, success, failure chan<- dispatchResult) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, n.cfg.DispatchTimeout.Value())
+	defer cancel()
+	logger := n.log.With(slog.F("msg_id", msg.ID), slog.F("method", msg.Method))
+
+	retryable, err := deliver(ctx, msg.ID)
+	if err != nil {
+		// Don't try to accumulate message responses if the context has been canceled.
+		//
+		// This message's lease will expire in the store and will be requeued.
+		// It's possible this will lead to a message being delivered more than once, and that is why Stop() is preferable
+		// instead of canceling the context.
+		//
+		// In the case of backpressure (i.e. the success/failure channels are full because the database is slow),
+		// we can't append any more updates to the channels otherwise this, too, will block.
+		if xerrors.Is(err, context.Canceled) {
+			return err
+		}
+
+		select {
+		case <-ctx.Done():
+			logger.Warn(context.Background(), "cannot record dispatch failure result", slog.Error(ctx.Err()))
+			return ctx.Err()
+		default:
+			logger.Warn(ctx, "message dispatch failed", slog.Error(err))
+			failure <- newFailedDispatch(n.id, msg.ID, err, retryable)
```

Comment:
This is still racy.  What you want is

```
select {
    case <-ctx.Done():
        return ctx.Err()
    case failure <- newFailedDispatch(n.id, msg.ID, err, retryable):
}
```

Reply:
It's true that `default` only executes if it is the only case that can succeed, but then writing to `failure` is a TOCTOU race.